### ISSUE TITLE
CupertinoPageRoute barrierColor is visible on a Scaffold beneath a transparent Widget

### ISSUE THREAD

**Issue by daveshirman** [flutter/flutter#105738]
## Steps to Reproduce
Please check [the DartPad here.](https://dartpad.dev/?id=95e0d2158d518965a8aebb817c79b50c)

## Notes: 
- I've determined that this issue is the Navigator object wrapping my `_page` property in the above DartPad.
- This exact code was rendering as expected on Flutter V2
- Upgraded to V3 and the bug is present.
- There is a comment on [this StackOverflow](https://stackoverflow.com/questions/72563512/flutter-why-is-navigator-changing-colour-of-scaffold-beneath-transparent-widget) question by someone who ran it on V2 and said it was fine.

**Note:** Bug result is same on iOS 15 and Android 12 physical devices
 
## To see the wrong/right render:

1. Load the above DartPad.
2. Toggle L.156 commented / uncommented 
3. Run
4. Observe that the render of the Scaffold blue colour is not consistent.

**Expected results:**

I expected: 
- there to be no discernible line or shading between the `BottomNavigationBar` and the `Scaffold` underneath
- the blue colour to be fully opaque, not somewhat "transparent" as it appears.

**Actual results:** 

- the blue colour of the underlying Scaffold is somewhat "transparent".

## Screenshots:

Bad:

<img width="776" alt="bad" src="https://user-images.githubusercontent.com/14280030/172940120-70c57086-1bd1-4923-8c4b-6f9c12eab155.png">

Good:

<img width="776" alt="good" src="https://user-images.githubusercontent.com/14280030/172940183-d65cd308-9970-4735-87fb-bdbb01847659.png">

```
Davids-MacBook-Pro:moodment daveshirman$ flutter doctor -v
[✓] Flutter (Channel stable, 3.0.1, on macOS 11.6.6 20G624 darwin-x64, locale en-GB)
    • Flutter version 3.0.1 at /Users/daveshirman/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fb57da5f94 (3 weeks ago), 2022-05-19 15:50:29 -0700
    • Engine revision caaafc5604
    • Dart version 2.17.1
    • DevTools version 2.12.2

[!] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/daveshirman/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.42.0

[✓] Connected device (1 available)
    • iPhone 13 mini (mobile) • FCD4F8E9-8E88-4C8C-B901-46D70A186315 • ios • com.apple.CoreSimulator.SimRuntime.iOS-15-2 (simulator)

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

**Comment by jason-simmons** [comment]
One possible fix would be adding `extendBody: true` to the `Scaffold` that contains the `bottomNavigationBar`

**Comment by daveshirman** [comment]
Thanks. So I knew already that adding that property would make it "seem"
fixed, but it wouldn't be in reality.

This is because the behaviour is clearly changed from V2 to V3.

If I make a Scaffold background transparent, I expect the widget behind to
show its full colour.

**Comment by danagbemava-nc** [comment]
Issue is reproducible on the latest stable and master, but not stable 2.10.5.

I played around with the code a bit, and it looks like this is due to the `CupertinoPageRoute` and not the `Navigator`. It seems like this is due to the `Navigator` because when you change from `CupertinoPageRoute` to `MaterialPageRoute` and hot reload, there are no visual changes, but when you hot-restart, you see that it works as expected. See the video below. 

Leaving this for further insight from the team.

<details>
<summary>video and screenshot</summary>

#### Screenshots

| stable 2.10.5 | flutter 3 | 
| -- | -- |
| <img src="https://user-images.githubusercontent.com/88313112/173036681-0ca3a593-219f-49c1-b92a-79687365428d.png" /> | <img src="https://user-images.githubusercontent.com/88313112/173037271-88d94df8-50a1-4458-a6d1-38c51d3d5b19.png" /> |

#### Video
https://user-images.githubusercontent.com/88313112/173036346-875107d7-fddd-4ff7-895c-d22748030461.mov

</details>


<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: _title,
      home: App(),
    );
  }
}

class App extends StatefulWidget {
  const App({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() => AppState();
}

class AppState extends State<App> {
  // this is static property so other widget throughout the app
  // can access it simply by AppState.currentTab

  static int currentTab = 0;

  // list tabs here
  final List<TabItem> tabs = [
    TabItem(
      tabName: "Home 1".toUpperCase(),
      page: const PageHome(),
    ),
    TabItem(
      tabName: "Home 2".toUpperCase(),
      page: const PageHome(),
    ),
  ];

  AppState() {
    // indexing is necessary for proper functionality
    // of determining which tab is active
    tabs.asMap().forEach((i, item) => item.setIndex(i));
  }

  @override
  void initState() {
    super.initState();
  }

  // sets current tab index
  // and update state
  void _selectTab(int index) {
    if (index == currentTab) {
      // pop to first route
      // if the user taps on the active tab
      tabs[index].key.currentState!.popUntil((route) => route.isFirst);
    } else {
      // update the state
      // in order to repaint
      if (mounted) {
        currentTab = index;
        _updateView();
      }
    }
  }

  void _updateView() {
    if (mounted) {
      setState(() {});
    }
  }

  void resetToHomeTab() {
    currentTab = 0;
  }

  @override
  Widget build(BuildContext context) {
    Widget body = Scaffold(
      backgroundColor: Colors.transparent,
      body: IndexedStack(
        index: currentTab,
        children: tabs.map((e) => e.page).toList(),
      ),
      bottomNavigationBar: BottomNavigation(
        onSelectTab: _selectTab,
        tabs: tabs,
      ),
    );

    // Background of the tab view
    Widget wrap = Container(
      width: double.infinity,
      height: double.infinity,
      color: Colors.blue,
      child: body,
    );

    // WillPopScope handle android back btn
    return WillPopScope(
      onWillPop: () async {
        final isFirstRouteInCurrentTab =
            !await tabs[currentTab].key.currentState!.maybePop();
        if (isFirstRouteInCurrentTab) {
          // if not on the 'main' tab
          if (currentTab != 0) {
            // select 'main' tab
            _selectTab(0);
            // back button handled by app
            return false;
          }
        }
        // let system handle back button if we're on the first route
        return isFirstRouteInCurrentTab;
      },
      // this is the base scaffold
      // don't put appbar in here otherwise you might end up
      // with multiple appbars on one screen
      // eventually breaking the app
      child: wrap,
    );
  }
}

class TabItem {
  final String tabName;
  final GlobalKey<NavigatorState> key = GlobalKey<NavigatorState>();
  int _index = 0;
  Widget? _page;
  TabItem({
    required this.tabName,
    required Widget page,
  }) {
    _page = page;
  }

  void setIndex(int i) {
    _index = i;
  }

  int getIndex() => _index;

  Widget get page {
    // If you toggle between the two return statements below, you can see the transparency
    // not working as expected when the page is wrapped in the Navigator

    // return _page!;

    return Visibility(
      visible: _index == AppState.currentTab,
      maintainState: true,
      // child: _page!,
      child: Navigator(
        key: key,
        onGenerateRoute: (routeSettings) {
          return CupertinoPageRoute(
            builder: (_) {
              return (_page != null) ? _page! : const SizedBox();
            },
          );
        },
      ),
    );
  }
}

class BottomNavigation extends StatelessWidget {
  const BottomNavigation({
    required this.onSelectTab,
    required this.tabs,
    Key? key
  }) : super(key: key);
  final ValueChanged<int> onSelectTab;
  final List<TabItem> tabs;

  @override
  Widget build(BuildContext context) {
    return BottomNavigationBar(
      elevation: 0,
      showSelectedLabels: true,
      showUnselectedLabels: true,
      backgroundColor: Colors.transparent,
      fixedColor: Colors.white,
      type: BottomNavigationBarType.fixed,
      items: tabs.map((e) => _buildItem(e.getIndex(), e.tabName)).toList(),
      onTap: (index) => onSelectTab(
        index,
      ),
      currentIndex: AppState.currentTab,
    );
  }

  BottomNavigationBarItem _buildItem(int index, String tabName) {
    return BottomNavigationBarItem(
      icon: const Padding(
        padding: EdgeInsets.only(bottom: 4),
        child: Icon(Icons.home),
      ),
      label: tabName,
    );
  }
}

class PageHome extends StatelessWidget {
  const PageHome({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      backgroundColor: Colors.transparent,
      body: Center(
        child: Text("Home page"),
      ),
    );
  }
}

```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel unknown, 2.10.5, on macOS 12.3.1 21E258 darwin-arm, locale en-GB)
    • Flutter version 2.10.5 at /Users/nexus/dev/sdks/flutter_rc/flutter
    • Upstream repository unknown
    • Framework revision 5464c5bac7 (8 weeks ago), 2022-04-18 09:55:37 -0700
    • Engine revision 57d3bac3dd
    • Dart version 2.16.2
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-32, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.42.0

[✓] Connected device (3 available)
    • iPhone 13 (mobile) • DBB48503-41CE-4969-9043-3261726E438A • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-4 (simulator)
    • macOS (desktop)    • macos                                • darwin-arm64   • macOS 12.3.1 21E258 darwin-arm
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 102.0.5005.61
    ! Error: Nexus is not connected. Xcode will continue when Nexus is connected. (code -13)

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```
[✓] Flutter (Channel stable, 3.0.1, on macOS 12.3.1 21E258 darwin-arm, locale en-GB)
    • Flutter version 3.0.1 at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fb57da5f94 (3 weeks ago), 2022-05-19 15:50:29 -0700
    • Engine revision caaafc5604
    • Dart version 2.17.1
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-32, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.42.0

[✓] Connected device (3 available)
    • iPhone 13 (mobile) • DBB48503-41CE-4969-9043-3261726E438A • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-4 (simulator)
    • macOS (desktop)    • macos                                • darwin-arm64   • macOS 12.3.1 21E258 darwin-arm
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 102.0.5005.61
    ! Error: Nexus is not connected. Xcode will continue when Nexus is connected. (code -13)

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```
[✓] Flutter (Channel master, 3.1.0-0.0.pre.1211, on macOS 12.3.1 21E258 darwin-arm, locale en-GB)
    • Flutter version 3.1.0-0.0.pre.1211 at /Users/nexus/dev/sdks/flutters
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 9d46e7800f (6 hours ago), 2022-06-09 23:43:08 -0400
    • Engine revision a4cdb1e232
    • Dart version 2.18.0 (build 2.18.0-174.0.dev)
    • DevTools version 2.14.0

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-32, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.42.0

[✓] Connected device (3 available)
    • iPhone 13 (mobile) • DBB48503-41CE-4969-9043-3261726E438A • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-4 (simulator)
    • macOS (desktop)    • macos                                • darwin-arm64   • macOS 12.3.1 21E258 darwin-arm
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 102.0.5005.61
    ! Error: Nexus is not connected. Xcode will continue when Nexus is connected. (code -13)

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>

**Comment by bleroux** [comment]
The "somewhat transparent" color visible here is the barrierColor introduced in https://github.com/flutter/flutter/pull/95537
I'm not sure why it is visible here maybe @willlockwood knows the answer?

**Comment by dkwingsmt** [comment]
@domph Hi! I don't think there's any update on this issue. Although I'd like to explain the "Cupertino TODO" that you mentioned.

I understand that you (and some other people) have started to notice the "Cupertino Refresh" github project and are getting excited by the improvement on Cupertino boosted by this push. That being said, the github project is _not_ meant to contain all Cupertino issues, since it can already be done that by filtering the Github issue list by the Cupertino label. 

There's a certain criteria of whether an issue is listed in this Github project, which is only based the work priority of me and a few of my colleagues and not based on importance. We had to reluctantly exclude a lot of obviously important issues from being listed there (such as the [calendar picker mode](https://github.com/flutter/flutter/issues/63696), but with the bandwidth limitation, listing too many issues there doesn't mean we're going to finish all of them. 

The bottom line is, there are many other work going on for Cupertino besides the "Cupertino Refresh“ Github project from both the Google Flutter team or the general contributor community. If someone is willing to investigate this issue and submit a PR, we'll be very happy to review it!
